ddprof-lib/src/main/cpp/callTraceStorage.cpp

```diff
@@ -8,195 +8,390 @@
 #include "counters.h"
 #include "os_dd.h"
 #include "common.h"
+#include "thread.h"
 #include "vmEntry.h" // For BCI_ERROR constant
 #include "arch_dd.h" // For LP64_ONLY macro and COMMA macro
+#include "criticalSection.h" // For table swap critical sections
 #include <string.h>
 #include <atomic>
 
+// HazardPointer static members
+std::atomic<CallTraceHashTable*> HazardPointer::global_hazard_list[HazardPointer::MAX_THREADS];
+
+// HazardPointer implementation
+int HazardPointer::getThreadHazardSlot() {
+    // Signal-safe: use thread ID hash to assign slots
+    // This avoids thread_local allocation issues in signal handlers
+    std::thread::id tid = std::this_thread::get_id();
+
+    // Apply Knuth multiplicative hash directly to thread ID hash
+    size_t hash = std::hash<std::thread::id>{}(tid) * KNUTH_MULTIPLICATIVE_CONSTANT;
+
+    // Use high bits for better distribution (shift right to get top bits)
+    return static_cast<int>((hash >> (sizeof(size_t) * 8 - 8)) % MAX_THREADS);
+}
+
+HazardPointer::HazardPointer(CallTraceHashTable* resource) : active_(true) {
+    // Get thread hazard slot using signal-safe method
+    my_slot_ = getThreadHazardSlot();
```

Comment:
What happens if there is collision?

Reply:
🤦 Right .. for CPU and wallclock, 256 slots should be quite enough. 
But for the JVMTI based samples we can have many more threads than the CPU cores.

I have expanded the hazard slot table, added bounded probing for conflict resolution and if we exhaust the table, we would just drop the trace instead of creating crash hazard.

The hazard slot table is now taking 64KiB which seems like a very small memory overhead we would pay for having pretty stable hazard pointer system.